proposals/0000-unstable.rst

```diff
@@ -0,0 +1,306 @@
+``-XUnstable``
+==============
+
+.. author:: Richard Eisenberg
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/524>`_.
+.. sectnum::
+.. contents::
+
+There are frequently new features in GHC that the GHC development team know are subject
+to change. Though we try to advertise features as unstable in e.g. the user manual and
+in release notes, we recognize it is impractical to expect every GHC user to read these
+resources.
+
+This proposal introduces a new extension ``-XUnstable``. Unstable features
+will be available only when ``-XUnstable`` is specified. Unstable features might
+change without notice, without a migration plan, and without a proposal, though
+notice of any changes would appear in release notes. In addition, we might choose
+to use this proposal process about unstable features, if collecting community
+feedback would be helpful.
+
+Moving
+an unstable feature out from the unstable umbrella would require a proposal.
+
+This proposal marks importing any module from ``ghc-prim`` as requiring ``-XUnstable``,
+though we expect to expand the set of features affected by ``-XUnstable`` over time.
+
+This proposal also includes a separate extension, ``-XStable``, which prevents
+``-XUnstable`` from being used. A project's ``.cabal`` file might then specify
+``-XStable`` to prevent any of the project's modules from using ``-XUnstable``.
+There is no ``-XNoStable``, unlike most other extensions. This ``-XStable`` component
+is optional and may be removed from this proposal without affecting the
+utility of ``-XUnstable``.
+
+Credit: This idea came up in a conversation between Richard Eisenberg and Simon
+Peyton Jones, though Richard fleshed out the details and wrote this proposal.
+
+Motivation
+----------
+
+``-XUnstable``
+~~~~~~~~~~~~~~
+
+It is frequently the case that the GHC developers know a feature is likely to
+change soon (or not-so-soon), and yet it is hard to spread this knowledge to
+the wider Haskell community. Examples:
+
+* For years, we GHC developers knew that ``-XImpredicativeTypes`` was more of
+  a bug than a feature. It had no specification, and any code that type-checked
+  because of it was simply lucky. The feature was in essence unstable. Since
```

Comment:
Did we know that even already when `-XImpredicativeTypes` was fist introduced, or was that hindsight? This only counts as motivation in the former case, does it?